stdlib/format.ml

```diff
@@ -441,80 +388,62 @@ let format_pp_token state size = function
       | Pp_vbox -> break_new_line state off width
       | Pp_hbox -> break_same_line state n
       end
-    | [] -> () (* No open box. *)
+    | None -> () (* No open box. *)
     end
 
    | Pp_open_tag tag_name ->
      let marker = state.pp_mark_open_tag tag_name in
      pp_output_string state marker;
-     state.pp_mark_stack <- tag_name :: state.pp_mark_stack
+     Stack.push tag_name state.pp_mark_stack
 
    | Pp_close_tag ->
-     begin match state.pp_mark_stack with
-     | tag_name :: tags ->
+     begin match Stack.pop_opt state.pp_mark_stack with
+     | Some tag_name ->
        let marker = state.pp_mark_close_tag tag_name in
        pp_output_string state marker;
-       state.pp_mark_stack <- tags
```

Comment:
Here, with the old version of the code, the mark was only removed from the state *after* `pp_output_string` had run. Is it possible that this would affect the behavior? (I'm not familiar with the codebase.)

Reply:
As I was drafting a response to you about how this would never affect the behavior, I actually… found a case where it affects the behavior.

`pp_output_string` delegates to `state.pp_out_string`. Built-in formatters set `state.pp_out_string` to `Pervasives.output_substring` or `Buffer.add_substring`. However, it could be set to anything, because it can be user-provided (by calling `pp_make_formatter`, for example). So if the user sets it to something that calls, for example, `pp_open_tag` on the same formatter, then:

* In the old code, the open tag would have been lost, overriden by the `<- tags`.
* In the new code, the open tag would be there, but handled before the close tag.

Neither of these make any sense, because it is such a contrived example.

In general, calling "porcelain" functions like `pp_open_tag` inside "plumbing" functions like "pp_output_string" would definitely wreak havoc in the internal state of the formatter. So I consider leaving this as it is.

Reply:
Ok. I'm convinced by the explanation. In general I buy the idea that popping early probably preserves the structure invariants better than popping late.